src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ModuleInitializerListNode.cs

```diff
@@ -0,0 +1,225 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+
+using Internal.Text;
+using Internal.TypeSystem;
+using Internal.TypeSystem.Ecma;
+
+namespace ILCompiler.DependencyAnalysis
+{
+    internal class ModuleInitializerListNode : ObjectNode, ISymbolDefinitionNode
+    {
+        private readonly ObjectAndOffsetSymbolNode _endSymbol;
+
+        public ModuleInitializerListNode()
+        {
+            _endSymbol = new ObjectAndOffsetSymbolNode(this, 0, "__module_initializers_End", true);
+        }
+
+        public ISymbolNode EndSymbol => _endSymbol;
+
+        public void AppendMangledName(NameMangler nameMangler, Utf8StringBuilder sb)
+        {
+            sb.Append(nameMangler.CompilationUnitPrefix).Append("__module_initializers");
+        }
+
+        public int Offset => 0;
+
+        public override bool IsShareable => false;
+
+        public override ObjectNodeSection Section => ObjectNodeSection.ReadOnlyDataSection;
+
+        public override bool StaticDependenciesAreComputed => true;
+
+        protected override string GetName(NodeFactory factory) => this.GetMangledName(factory.NameMangler);
+
+        public override ObjectData GetData(NodeFactory factory, bool relocsOnly = false)
+        {
+            // This is a summary node that doesn't introduce dependencies.
+            if (relocsOnly)
+                return new ObjectData(Array.Empty<byte>(), Array.Empty<Relocation>(), 1, new ISymbolDefinitionNode[] { this });
+
+            
+            var modulesWithCctor = new HashSet<ModuleDesc>();
+
+            foreach (var methodNode in factory.MetadataManager.GetCompiledMethodBodies())
+            {
+                if (methodNode.Method.IsStaticConstructor && methodNode.Method.OwningType is MetadataType mdType
+                    && mdType.IsModuleType)
+                {
+                    modulesWithCctor.Add(mdType.Module);
+                }
+            }
+
+            // We have a list of modules with a class constructor.
+            // Do a topological sort based on the assembly references of each module.
+            // This is an approximation that tries to deal with module initializers that might have
+            // dependencies on each other. The spec doesn't guarantee any ordering so this is best effort.
+            List<ModuleDesc> sortedModules = new List<ModuleDesc>();
+
+            var graphFactory = new ModuleGraphFactory();
+
+            // This is a list because we want to keep a stable sort order.
```

Comment:
This is populated from `HashSet<modulesWithCctor>`. Do we need to also need to do something to make the HashSet order stable?

Reply:
Yes, good catch. Made it a list.